openvdb_points/tools/AttributeArrayString.h

```diff
@@ -0,0 +1,167 @@
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file AttributeArrayString.h
+///
+/// @authors Dan Bailey
+///
+/// @brief  Attribute array storage for string data using Descriptor Metadata.
+///
+
+
+#ifndef OPENVDB_TOOLS_ATTRIBUTE_ARRAY_STRING_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_ATTRIBUTE_ARRAY_STRING_HAS_BEEN_INCLUDED
+
+#include <openvdb_points/tools/AttributeArray.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+////////////////////////////////////////
+
+
+template <typename Codec_ = NullAttributeCodec<StringType> >
+class StringTypedAttributeArray : public TypedAttributeArray<StringType, Codec_>
+{
+public:
+    typedef StringType                  ValueType;
+    typedef Codec_                      Codec;
+    typedef typename Codec::StorageType StorageType;
+
+    /// Default constructor, always constructs a uniform attribute.
+    explicit StringTypedAttributeArray(  size_t n = 1,
+                                    const ValueType& uniformValue = zeroVal<ValueType>())
+        : TypedAttributeArray<StringType, Codec_>(n, uniformValue) { this->setString(true); }
+
+    /// Deep copy constructor (optionally decompress during copy).
+    StringTypedAttributeArray(const StringTypedAttributeArray& array,
+                        const bool decompress = false)
+        : TypedAttributeArray<StringType, Codec_>(array, decompress) { this->setString(true); }
+
+}; // class StringTypedAttributeArray
+
+
+// Convenience typedef when using with no codec
+typedef StringTypedAttributeArray<>     StringAttributeArray;
+
+
+////////////////////////////////////////
+
+
+class StringAttributeHandle
+{
+public:
+    typedef boost::shared_ptr<StringAttributeHandle> Ptr;
+
+    static Ptr create(const AttributeArray& array, const MetaMap& metadata, const bool preserveCompression = true);
+
+    StringAttributeHandle(  const AttributeArray& array,
+                            const MetaMap& metadata,
+                            const bool preserveCompression = true);
```

Comment:
Interesting how you're storing the string data on metadata for every descriptor. Can we please discuss this, have a few questions.